### ISSUE TITLE
[KED-1635] Improve DataSetNotFoundError: messages for MemoryDataSets

### ISSUE THREAD

**Issue by WaylonWalker** [quantumblacklabs/kedro#336]
## Description
I'm always frustrated when I get a `DataSetNotFoundError:` for MemoryDataSets.  Would it be possible to look at pipeline.all_inputs to see if the dataset is part of the pipeline, but not in the catalog, then provide instructions to get the data loaded?

## Context
I like chaining many small functions together and Its not always necessary to save every step along the way, but I find myself creating a catalog entry for each node so that its not confusing for others later on.

## Possible Implementation
I am not sure if it is possible to access the MemoryDataSet after a pipeline run.  But if the error message could find the shortest distance between a MemoryDataSet and a catalog entry then provide some instruction to get there.

**potential error message**
``` diff
 ---------------------------------------------------------------------------
DataSetNotFoundError                      Traceback (most recent call last)
<ipython-input-22-0b55c03ca6c5> in <module>
 ----> 1 catalog.load("C")
/path/to/data_catalog.py in load(self, name, version)
    347         if name not in self._data_sets:
    348             raise DataSetNotFoundError(
--> 349                 "DataSet '{}' not found in the catalog".format(name)
    350             )
    351
- DataSetNotFoundError: DataSet 'C' not found in the catalog
+ DataSetNotFoundError: DataSet 'C' is a MemoryDatasSet
+ run(pipeline.from_inputs('B') - pipeline.to_outputs('C'), keep_memory_data_set_after_run=True)
+ catalog.load('C')
```

## Possible Alternatives
Let me know your thoughts or if anything like this would be possible.

**Comment by yetudada** [comment]
We'll also have a look into this one. Thanks for raising this! 🎉

**Comment by lorenabalan** [comment]
Hi @WaylonWalker , thank you the issue! Unfortunately we decided to not move forward with this.

We'd like to keep the DataCatalog and the Pipeline independent from each other, for users to still be able to pick and choose bits from the library to use, and this change would make the two tightly coupled. We also can't guarantee that it's a `MemoryDataSet`. It is the responsibility of the runner to create a default dataset to be used for data that's not in the catalog, so a custom runner could rely on a different default DataSet and it would go out of sync with the error message. 

Having said that, we did discuss internally that we would like to take a closer look at the interactive workflow, which your issue seems to be part of, and think of it more broadly and properly address it. We'll make sure to think of your feedback within that scope!
